### PR TITLE
shuf: improve performance

### PR DESCRIPTION
**Pull request by Garfield96**
This PR improves the performance of `shuf` by replacing the current implementation of splitting input by a given separator with a version using the memchr crate. This crate was not used in `shuf` before, but it is already used in e.g. sort. Considering the benchmark results (see below), adding a new dependency to `shuf` is justified.

### Description

Generate input data (50,000,000 lines, 3.6 GiB):
```
cat /dev/random | base64 | fold | head -n 50000000 > input.txt
```
Run `shuf` with [hyperfine](https://github.com/sharkdp/hyperfine):
```
hyperfine -w 3 "target/release/coreutils shuf ../testdata/input.txt > /dev/null"
```
### Results

The input data was stored in tmpfs to avoid distortions from IO.

| Version  |  Time |
|---|---|
| Old version |  6.982 s ±  0.033 s |
| Improved version |  5.234 s ±  0.046 s |
| GNU shuf (v8.32) | 9.810 s ±  0.019 s |

The improved version is 25% faster than the previous version. Compared to the GNU implementation, `shuf` now is almost twice as fast.

### CODE REVIEW

**Comment by sylvestre** [comment]
Cool stuff. Could you please create a src/uu/shuf/BENCHMARKING.md  file ?
(see src/uu/*/BENCHMARKING.md  for example)

**Comment by sylvestre** [comment]
Some more benchmarks:
```
$ wget -O shakespeare.txt https://www.gutenberg.org/files/100/100-0.txt
$ hyperfine   -w 3 "./target/release/coreutils shuf shakespeare.txt > /dev/null" "shuf shakespeare.txt > /dev/null"              
Benchmark 1: ./target/release/coreutils shuf shakespeare.txt > /dev/null
  Time (mean ± σ):      22.1 ms ±   4.6 ms    [User: 15.3 ms, System: 6.8 ms]
  Range (min … max):    10.7 ms …  28.4 ms    114 runs
 
Benchmark 2: shuf shakespeare.txt > /dev/null
  Time (mean ± σ):      27.8 ms ±   7.4 ms    [User: 22.1 ms, System: 5.7 ms]
  Range (min … max):    14.5 ms …  42.1 ms    72 runs
 
Summary
  './target/release/coreutils shuf shakespeare.txt > /dev/null' ran
    1.26 ± 0.42 times faster than 'shuf shakespeare.txt > /dev/null'
```

=> faster here

But slower with:
```
$ wget https://www.gutenberg.org/files/30613/30613-0.txt -O odyssey.txt
$ hyperfine --shell=none  -w 3 "./target/release/coreutils shuf odyssey.txt " "shuf odyssey.txt "
Benchmark 1: ./target/release/coreutils shuf odyssey.txt 
  Time (mean ± σ):       3.4 ms ±   0.3 ms    [User: 1.6 ms, System: 1.6 ms]
  Range (min … max):     2.0 ms …   3.8 ms    866 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Benchmark 2: shuf odyssey.txt 
  Time (mean ± σ):       2.2 ms ±   0.3 ms    [User: 1.5 ms, System: 0.5 ms]
  Range (min … max):     1.1 ms …   3.4 ms    1209 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Summary
  'shuf odyssey.txt ' ran
    1.53 ± 0.24 times faster than './target/release/coreutils shuf odyssey.txt '
```

**Comment by anastygnome** [comment]
`bash

❯ hyperfine --shell=none  -w 3 "./target/release/coreutils shuf /tmp/shakespeare.txt " "shuf /tmp/shakespeare.txt "

Benchmark 1: ./target/release/coreutils shuf /tmp/shakespeare.txt 
  Time (mean ± σ):      15.9 ms ±   5.5 ms    [User: 11.3 ms, System: 4.4 ms]
  Range (min … max):    10.1 ms …  24.4 ms    281 runs
 
Benchmark 2: shuf /tmp/shakespeare.txt 
  Time (mean ± σ):      21.8 ms ±   6.7 ms    [User: 17.2 ms, System: 4.4 ms]
  Range (min … max):    14.8 ms …  36.1 ms    193 runs
 
Summary
  './target/release/coreutils shuf /tmp/shakespeare.txt ' ran
    1.37 ± 0.64 times faster than 'shuf /tmp/shakespeare.txt '

❯ hyperfine --shell=none  -w 3 "./target/release/coreutils shuf /tmp/odyssey.txt " "shuf /tmp/odyssey.txt "

Benchmark 1: ./target/release/coreutils shuf /tmp/odyssey.txt 
  Time (mean ± σ):       1.4 ms ±   0.5 ms    [User: 0.9 ms, System: 0.4 ms]
  Range (min … max):     0.8 ms …   2.5 ms    1527 runs
 
Benchmark 2: shuf /tmp/odyssey.txt 
  Time (mean ± σ):       1.4 ms ±   0.5 ms    [User: 0.9 ms, System: 0.4 ms]
  Range (min … max):     0.8 ms …   2.4 ms    1729 runs
 
Summary
  'shuf /tmp/odyssey.txt ' ran
    1.02 ± 0.53 times faster than './target/release/coreutils shuf /tmp/odyssey.txt '

`

@sylvestre nice I/O benchmark of yours there ;)

You should put the file in tmpfs to negate this issue (in /tmp/ in my case)

**Comment by sylvestre** [comment]
Cool :)
Just create the benchmark doc and it can land!

**Comment by anastygnome** [comment]
> Cool :) Just create the benchmark doc and it can land!

I'm not the PR guy X) 

But I'll try :)

**Comment by sylvestre** [comment]
oups, sorry :)
Please let @Garfield96 do it :)

**Comment by anastygnome** [comment]
> oups, sorry :) Please let @Garfield96 do it :)

On a side note, please look at the following markdown syntax for CLI output 
```markdown

```console
$ whoami 
anastygnome 
<3x ` to END>
```
yields 

```console
$ whoami 
anastygnome 
```

**Comment by Garfield96** [comment]
Thanks, @sylvestre, and @anastygnome for doing additional benchmarks. I also tested `https://www.gutenberg.org/files/30613/30613-0.txt` and I can confirm that the GNU version in some runs showed a slightly better performance. Since this input is very small, applying the shuffling does not dominate the execution time. Instead, loading the binary becomes the bottleneck. While the combined build of coreutils always produced worse results on my system (1.2ms compared to 0.9ms for GNU shuf), building `shuf` as a standalone binary solved the problem (runtimes were equal).

I extended the description in `BENCHMARKING.md`, hence this PR should be complete.

File: src/uu/shuf/BENCHMARKING.md [hunk 1]
```diff
@@ -1,26 +1,51 @@
...
-You can compare with another branch by compiling on that branch and than
+You can compare with another branch by compiling on that branch and then
 renaming the executable from `shuf` to `shuf.old`.
 
+## Generate sample data
+
+Sample input can be generated using `/dev/random`:
+
+```shell
+cat /dev/random | base64 | fold | head -n 50000000 > input.txt
```
Comment (sylvestre) [inline src/uu/shuf/BENCHMARKING.md:17]:
```suggestion
wget -O input.txt https://www.gutenberg.org/files/100/100-0.txt
```

**Pull request merged**